### PR TITLE
Feature/wa 362 wallet service

### PR DESCRIPTION
**Pull request by DmitryBespalov**
- `MultisigWalletApplication`
  - Registry
  - Account
  - Implemented methods using domain model
- `MultisigWalletDomainModel`
  - Wallet now has same states as listed in wallet application service
- `MultisigWalletImplementations`
  - Added AccountRepository
- Fixed UI tests
- Implemented TODOs (logger)
- Moved Logger protocol and mock out to Common module
- Created CommonImplementations module with logger's implementation moved there.

### CODE REVIEW

File: Common/CommonImplementations/logger/LogFormatter.swift [hunk 1]
```diff
@@ -16,7 +16,7 @@ import Foundation
 final class LogFormatter {
 
     static let defaultDateFormat = "yyyy-MM-dd hh:mm:ss.SSSSSS"
-    static let defaultMessageFormat = "%t [%l] %f:%n %m: %s"
+    static let defaultMessageFormat = "%t [%l] %f:%n %m: %s %e"
```
Comment (DmitryBespalov) [inline Common/CommonImplementations/logger/LogFormatter.swift:19]: error was not printed in logs before

File: SafeUIKit/SafeAppUI/ErrorHandler.swift [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    private init() {}
+
+    public static func showFatalError(message: String = Strings.fatalErrorMessage, log: String, error: Error?) {
+        ApplicationServiceRegistry.logger.fatal(log, error: error)
+        instance.showError(message: message, log: log, error: error) {
+            fatalError(message + "; " + log)
+        }
+    }
+
+    public static func showError(message: String = Strings.errorMessage, log: String, error: Error?) {
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeAppUI/ErrorHandler.swift:30]: This is now just showing error, without terminating the app

File: MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift [hunk 3]
```diff
@@ -26,50 +28,253 @@ public class WalletApplicationService {
...
+                } else {
+                    return .accountFunded
+                }
+            case .deploymentAcceptedByBlockchain:
+                return .deploymentAcceptedByBlockchain
+            case .deploymentSuccess:
+                return .deploymentSuccess
+            case .deploymentFailed:
+                return .deploymentFailed
+            case .readyToUse:
```
Comment (sche) [inline MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift:74]: do we need it?

File: MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift [hunk 4]
```diff
@@ -26,50 +28,253 @@ public class WalletApplicationService {
...
+
+    private func observeWalletStateWhile(_ closure: () throws -> Void) rethrows {
+        let startState = selectedWalletState
+        try closure()
+        let endState = selectedWalletState
+        if startState != endState {
+            notifyStatusUpdate()
+        }
+    }
+    // MARK: - Owners
```
Comment (sche) [inline MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift]: please add empty line

File: MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift [hunk 5]
```diff
@@ -26,50 +28,253 @@ public class WalletApplicationService {
...
+
+    public func updateMinimumFunding(account token: String, amount: Int) throws {
+        try assertCanChangeAccount()
+        try mutateAccount(token: token) { account in
+            account.updateMinimumTransactionAmount(amount)
+        }
+    }
+
+    private func assertCanChangeAccount() throws {
+        try assertTrue(WalletApplicationService.validAccountUpdateStates.contains(selectedWalletState),
```
Comment (sche) [inline MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift:236]: [.deploymentAcceptedByBlockchain, .deploymentSuccess] should not be in validAccountUpdateStates

Comment (DmitryBespalov) [reply MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift:236]: why not? the balance update can happen any time in blockchain after we know the account's address.

File: MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift (hunk 3 above)
Comment (DmitryBespalov) [reply MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift:74]: let's keep it for now, I think some other states will go away when we'll have working ethereum solution.

File: MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift [hunk 6]
```diff
@@ -26,50 +28,253 @@ public class WalletApplicationService {
...
+    private func findSelectedWallet() throws -> Wallet {
+        guard let portfolio = try DomainRegistry.portfolioRepository.portfolio(),
+            let walletID = portfolio.selectedWallet,
+            let wallet = try DomainRegistry.walletRepository.findByID(walletID) else {
+                throw Error.selectedWalletNotFound
+        }
+        return wallet
+    }
+
+    private func observeWalletStateWhile(_ closure: () throws -> Void) rethrows {
```
Comment (sche) [inline MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift]: How about withNotificationOnStateChange or smth similar? Current name confuses me.

Comment (DmitryBespalov) [reply MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift]: `notifyWalletStatusChangesWhile`?

File: MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift [hunk 7]
```diff
@@ -26,50 +28,253 @@ public class WalletApplicationService {
...
+    private func findAccount(_ token: String) throws -> Account {
+        let wallet = try findSelectedWallet()
+        guard let account = try DomainRegistry.accountRepository.find(id: AccountID(token: token),
+                                                                      walletID: wallet.id) else {
+                                                                        throw Error.accountNotFound
+        }
+        return account
+    }
+
+    // MARK: - Wallet status update subsribing
```
Comment (sche) [inline MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift]: subsCribing

File: MultisigWallet/MultisigWalletApplication/WalletApplicationServiceTests.swift [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+import XCTest
+@testable import MultisigWalletApplication
+import MultisigWalletImplementations
+import MultisigWalletDomainModel
+import Common
+
+class WalletApplicationServiceTests: XCTestCase {
+
+    let walletRepository = InMemoryWalletRepository()
```
Comment (sche) [inline MultisigWallet/MultisigWalletApplication/WalletApplicationServiceTests.swift:13]: How about moving them to WalletApplicationTestCase?

File: MultisigWallet/MultisigWalletApplication/WalletApplicationServiceTests.swift [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+    func test_whenCreatingFirstWallet_thenCanObserveStatusUpdate() {
+        var updated = false
+        _ = service.subscribe {
+            updated = true
+        }
+        givenDraftWallet()
+        XCTAssertTrue(updated)
+    }
+
+    func test_whenWalletIsReady_thenHasReadyState() throws {
```
Comment (sche) [inline MultisigWallet/MultisigWalletApplication/WalletApplicationServiceTests.swift:198]: looks like duplicate with test_fullCycle

File: MultisigWallet/MultisigWalletApplication/WalletApplicationServiceTests.swift (hunk 8 above)
Comment (DmitryBespalov) [reply MultisigWallet/MultisigWalletApplication/WalletApplicationServiceTests.swift:13]: There's only 1 test file using these things, I'll keep it here

File: MultisigWallet/MultisigWalletApplication/WalletApplicationServiceTests.swift (hunk 9 above)
Comment (DmitryBespalov) [reply MultisigWallet/MultisigWalletApplication/WalletApplicationServiceTests.swift:198]: This uses the same sequence but tests another API - `hasReadyToUseWallet` variable in the service. The `test_fullCycle` tests `selectedWalletState`

**Review by sche — APPROVED** [review]

**Pull request merged**